pulsar-broker/src/main/java/org/apache/pulsar/client/impl/RawBatchConverter.java

```diff
@@ -135,7 +135,8 @@ public static Optional<RawMessage> rebatchMessage(RawMessage msg,
                                                       msg.getMessageIdData().getPartition(),
                                                       i);
                 if (!singleMessageMetadata.hasPartitionKey()) {
-                    if (retainNullKey) {
+                    // we may read compacted out message from compacted topic, which do not have partition key too.
+                    if (retainNullKey && !singleMessageMetadata.isCompactedOut()) {
```

Comment:
Can we move the `isCompactedOut` judgment outside?

Reply:
There is no need to move it out, because if one message is compacted out, it's `PartitionKey` in the metadata must be null, it is the subcase of the `!singleMessageMetadata.hasPartitionKey()`.

Reply:
done.